apoc.import.csv creates superflous nodes when relationship references don't exist

Test case, using the example from "Loading nodes and relationships":

`/tmp/persons.csv`:
```
:ID,name:STRING,speaks:STRING[]
1,John,"en,fr"
2,Jane,"en,de"
```

`/tmp/knows.csv`:
```
:START_ID,:END_ID,since:INT
1,2,2016
10,11,2014
11,12,2013
```
(note the 2 extra lines)

Command:
```
CALL apoc.import.csv(
  [{fileName: '/tmp/persons.csv', labels: ['Person']}],
  [{fileName: '/tmp/knows.csv', type: 'KNOWS'}],
  {stringIds: false}
)
```

The result creates 6 nodes - John and Jane are tripled:
```
╒═══════════════════════════════════════════════╕
│"n"                                            │
╞═══════════════════════════════════════════════╡
│{"name":"John","speaks":["en,fr"],"__csv_id":1}│
├───────────────────────────────────────────────┤
│{"name":"Jane","speaks":["en,de"],"__csv_id":2}│
├───────────────────────────────────────────────┤
│{"name":"John","speaks":["en,fr"],"__csv_id":1}│
├───────────────────────────────────────────────┤
│{"name":"Jane","speaks":["en,de"],"__csv_id":2}│
├───────────────────────────────────────────────┤
│{"name":"John","speaks":["en,fr"],"__csv_id":1}│
├───────────────────────────────────────────────┤
│{"name":"Jane","speaks":["en,de"],"__csv_id":2}│
└───────────────────────────────────────────────┘
```

This is the full response:
```
[
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 393,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "John",
          "speaks": [
            "en,fr"
          ],
          "__csv_id": {
            "low": 1,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  },
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 394,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "Jane",
          "speaks": [
            "en,de"
          ],
          "__csv_id": {
            "low": 2,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  },
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 414,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "John",
          "speaks": [
            "en,fr"
          ],
          "__csv_id": {
            "low": 1,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  },
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 415,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "Jane",
          "speaks": [
            "en,de"
          ],
          "__csv_id": {
            "low": 2,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  },
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 488,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "John",
          "speaks": [
            "en,fr"
          ],
          "__csv_id": {
            "low": 1,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  },
  {
    "keys": [
      "n"
    ],
    "length": 1,
    "_fields": [
      {
        "identity": {
          "low": 489,
          "high": 0
        },
        "labels": [
          "Person"
        ],
        "properties": {
          "name": "Jane",
          "speaks": [
            "en,de"
          ],
          "__csv_id": {
            "low": 2,
            "high": 0
          }
        }
      }
    ],
    "_fieldLookup": {
      "n": 0
    }
  }
]
```

I tried to prevent this by setting a constraint on `name`, but that doesn't prevent the nodes from being created: they will be created without a name and without a relationship. Related but likely a different bug.

@jexp sure - I did not notice this issue before.

@ArthurClemens thanks for the detailed error report.

I think the underlying issue is that this CSV loader was designed to work with _very clean data_, i.e. the assumption is that there are no missing nodes, no type conversion errors. If anything is wrong, the load crashes.

Now, I added the test case from https://github.com/neo4j-contrib/neo4j-apoc-procedures/issues/951#issue-369571828 in commit https://github.com/szarnyasg/neo4j-apoc-procedures/commit/c9ad03bd127400c59f651a887d2117615de83c10.

However, for me, it does not create duplicate nodes, but simply blows us spectacularly (which I'd say is the expected behaviour):

```console
org.neo4j.graphdb.QueryExecutionException: Failed to invoke procedure `apoc.import.csv`: Caused by: java.lang.IllegalStateException: Node for id space __CSV_DEFAULT_IDSPACE and id 10 not found

	at org.neo4j.kernel.impl.query.QueryExecutionKernelException.asUserException(QueryExecutionKernelException.java:35)
	at org.neo4j.kernel.impl.factory.ClassicCoreSPI.executeQuery(ClassicCoreSPI.java:83)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.execute(GraphDatabaseFacade.java:428)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.execute(GraphDatabaseFacade.java:411)
	at apoc.util.TestUtil.testResult(TestUtil.java:89)
	at apoc.util.TestUtil.testCall(TestUtil.java:38)
	at apoc.export.csv.ImportCsvTest.testNoDuplicationsCreated(ImportCsvTest.java:403)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
```

@ArthurClemens can you check my test case?

--- a/src/test/java/apoc/export/csv/ImportCsvTest.java
+++ b/src/test/java/apoc/export/csv/ImportCsvTest.java
@@ -63,6 +63,13 @@ public class ImportCsvTest {
                             "2|1|b|6\n"),
                     new AbstractMap.SimpleEntry<>("label", ":ID|:LABEL|name:STRING\n" +
                             "1|Student;Employee|John\n"),
+                    new AbstractMap.SimpleEntry<>("knows", ":START_ID,:END_ID,since:INT\n" +
+                            "1,2,2016\n" +
+                            "10,11,2014\n" +
+                            "11,12,2013"),
+                    new AbstractMap.SimpleEntry<>("persons", ":ID,name:STRING,speaks:STRING[]\n" +
+                            "1,John,\"en,fr\"\n" +
+                            "2,Jane,\"en,de\""),
                     new AbstractMap.SimpleEntry<>("quoted", "id:ID|:LABEL|name:STRING\n" +
                             "'1'|'Student:Employee'|'John'\n"),
                     new AbstractMap.SimpleEntry<>("rel-on-ids-idspaces", ":START_ID(Person)|:END_ID(Person)|since:INT\n" +
@@ -389,6 +396,23 @@ public class ImportCsvTest {
             );
         Assert.assertEquals("Jane 26 <6> John 25", result2.next().get("pair"));
         Assert.assertEquals("John 25 <3> Jane 26", result2.next().get("pair"));
+    }
+
+    @Test
+    public void testNoDuplicationsCreated() {
+        TestUtil.testCall(
+                db,
+                "CALL apoc.import.csv([{fileName: {nodeFile}, labels: ['Person']}], [{fileName: {relFile}, type: 'KNOWS'}], {config})",
+                map(
+                        "nodeFile", "file:/persons.csv",
+                        "relFile", "file:/knows.csv",
+                        "config", map("stringIds", false)
+                ),
+                (r) -> {
+                    assertEquals(2L, r.get("nodes"));
+                    assertEquals(1L, r.get("relationships"));
+                }
+        );
     }
 
 }

